File: Sliding window/LET_no159_Longest Substring with At Most Two Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        memo = collections.defaultdict(int)
        res_left = res_right = 0
        left = 0
        for right in range(len(s)):
            memo[s[right]] += 1
            while len(memo) > 2 and left + 1 < right:
                memo[s[left]] -= 1
                if memo[s[left]] == 0:
                    del memo[s[left]]
                left += 1
            
            if right - left > res_right - res_left:
                res_left, res_right = left, right
        
        return res_right - res_left + 1

File: Sliding window/LET_no159_Longest Substring with At Most Two Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        unordered_map<char, int> memo;
        int i = 0, j = 0, left = 0;
        for (int right = 0; right < s.size(); right++) {
            memo[s[right]]++;
            while (memo.size() > 2 && left + 1 < right) {
                memo[s[left]]--;
                if (memo[s[left]] == 0)
                    memo.erase(s[left]);
                left++;
            }

            if (right - left > j - i) {
                i = left;
                j = right;
            }
        }

        return j - i + 1;
    }
};